File: sdks/java/src/main/java/io/sentry/playground/controller/TransformController.java
package io.sentry.playground.controller;

import io.sentry.playground.dto.TransformRequest;
import io.sentry.playground.dto.TransformResponse;
import io.sentry.playground.service.TransformService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.lang.reflect.Method;
import java.lang.reflect.Field;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

@RestController
public class TransformController {

    @Autowired
    private TransformService transformService;

    @PostMapping("/transform")
    public ResponseEntity<TransformResponse> transform(@RequestBody TransformRequest request) {
        // Validate request
        if (request.getEvent() == null || request.getBeforeSendCode() == null ||
            request.getBeforeSendCode().trim().isEmpty()) {
            TransformResponse response = new TransformResponse(
                false,
                null,
                "Missing event or beforeSendCode",
                null
            );
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // Transform the event
        TransformService.TransformResult result = transformService.transform(
            request.getEvent(),
            request.getBeforeSendCode()
        );

        // Build response
        TransformResponse response = new TransformResponse(
            result.isSuccess(),
            result.getTransformedEvent(),
            result.getError(),
            result.getTraceback()
        );

        // Return appropriate status code
        if (result.isSuccess()) {
            return ResponseEntity.ok(response);
        } else if (result.getTraceback() != null) {
            // Runtime error
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        } else {
            // Compilation error or validation error
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }
    }

    @PostMapping("/validate-config")
    public ResponseEntity<Map<String, Object>> validateConfig(@RequestBody Map<String, String> request) {
        String configCode = request.get("configCode");
        if (configCode == null || configCode.trim().isEmpty()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of(
                "success", false,
                "error", "Missing configCode"
            ));
        }

        String sdkVersion = "unknown";
        try {
            Class<?> sentryClass = Class.forName("io.sentry.Sentry");
            // Try to get the SDK version
            try {
                Class<?> buildConfig = Class.forName("io.sentry.BuildConfig");
                Field versionField = buildConfig.getField("VERSION_NAME");
                sdkVersion = (String) versionField.get(null);
            } catch (Exception e) {
                sdkVersion = sentryClass.getPackage().getImplementationVersion() != null
                    ? sentryClass.getPackage().getImplementationVersion() : "unknown";
            }
        } catch (Exception e) {
            // ignore
        }

        List<String> warnings = new ArrayList<>();
        Map<String, Object> result = new HashMap<>();
        result.put("success", true);
        result.put("sdk", "java");
        result.put("sdkVersion", sdkVersion);
        result.put("resolvedOptions", new HashMap<>());
        result.put("recognizedKeys", new ArrayList<>());
        result.put("ignoredKeys", new ArrayList<>());

        // Capture System.err to detect warnings during init
        PrintStream originalErr = System.err;
        ByteArrayOutputStream capturedErr = new ByteArrayOutputStream();
        System.setErr(new PrintStream(capturedErr));

        try {
            // Use Groovy to execute the config code
            groovy.lang.GroovyShell shell = new groovy.lang.GroovyShell();
            shell.evaluate(configCode);
            result.put("initSucceeded", true);
        } catch (Exception e) {
            result.put("initSucceeded", false);
            result.put("error", e.getMessage());
        } finally {
            System.setErr(originalErr);
            String errOutput = capturedErr.toString().trim();
            if (!errOutput.isEmpty()) {
                for (String line : errOutput.split("\n")) {
                    if (!line.trim().isEmpty()) {
                        warnings.add(line.trim());
                    }
                }
            }
            // Close Sentry to release background threads and shutdown hooks
            try {
                io.sentry.Sentry.close();
            } catch (Exception e) {
                // ignore
            }
        }

        result.put("warnings", warnings);
        return ResponseEntity.ok(result);
    }

    @GetMapping("/introspect")
    public ResponseEntity<Map<String, Object>> introspect() {
        try {
            String sdkVersion = "unknown";
            List<Map<String, Object>> options = new ArrayList<>();

            try {
                Class<?> sentryOptionsClass = Class.forName("io.sentry.SentryOptions");
                sdkVersion = sentryOptionsClass.getPackage().getImplementationVersion() != null
                    ? sentryOptionsClass.getPackage().getImplementationVersion() : "unknown";

                // Use reflection to discover setter methods (which represent configurable options)
                Method[] methods = sentryOptionsClass.getMethods();
                for (Method method : methods) {
                    String name = method.getName();
                    if (!name.startsWith("set") || method.getParameterCount() != 1) continue;

                    String optionName = name.substring(3);
                    // Convert PascalCase to camelCase
                    String canonicalKey = Character.toLowerCase(optionName.charAt(0)) + optionName.substring(1);

                    Class<?> paramType = method.getParameterTypes()[0];
                    String typeStr;
                    if (paramType == String.class) typeStr = "string";
                    else if (paramType == boolean.class || paramType == Boolean.class) typeStr = "boolean";
                    else if (paramType == double.class || paramType == Double.class || paramType == float.class || paramType == Float.class) typeStr = "float";
                    else if (paramType == int.class || paramType == Integer.class || paramType == long.class || paramType == Long.class) typeStr = "integer";
                    else if (paramType.isArray() || List.class.isAssignableFrom(paramType)) typeStr = "array";
                    else typeStr = paramType.getSimpleName().toLowerCase();

                    Map<String, Object> opt = new HashMap<>();
                    opt.put("key", optionName);
                    opt.put("canonicalKey", canonicalKey);
                    opt.put("type", typeStr);
                    opt.put("required", canonicalKey.equals("dsn"));
                    opt.put("default", null);
                    opt.put("description", "");
                    options.add(opt);
                }
            } catch (Exception e) {
                // SentryOptions not on classpath - return empty
            }

            Map<String, Object> result = new HashMap<>();
            result.put("sdk", "java");
            result.put("sdkVersion", sdkVersion);
            result.put("sdkPackage", "io.sentry:sentry");
            result.put("source", "reflection");
            result.put("options", options);
            result.put("timestamp", java.time.Instant.now().toString());

            return ResponseEntity.ok(result);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of(
                "success", false,
                "error", "Introspection service error: " + e.getMessage()
            ));
        }
    }

    @GetMapping("/health")
    public ResponseEntity<Map<String, String>> health() {
        return ResponseEntity.ok(Map.of(
            "status", "healthy",
            "sdk", "java"
        ));
    }
}

File: sdks/android/src/main/kotlin/io/sentry/playground/controller/TransformController.kt
package io.sentry.playground.controller

import io.sentry.playground.dto.TransformRequest
import io.sentry.playground.dto.TransformResponse
import io.sentry.playground.service.TransformService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
class TransformController(private val transformService: TransformService) {

    @PostMapping("/transform")
    fun transform(@RequestBody request: TransformRequest): ResponseEntity<TransformResponse> {
        // Validate request
        if (request.beforeSendCode.trim().isEmpty()) {
            val response = TransformResponse(
                success = false,
                error = "Missing event or beforeSendCode",
                traceback = null
            )
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response)
        }

        // Transform the event
        val result = transformService.transform(
            request.event,
            request.beforeSendCode
        )

        // Build response
        val response = TransformResponse(
            success = result.success,
            transformedEvent = result.transformedEvent,
            error = result.error,
            traceback = result.traceback
        )

        // Return appropriate status code
        return when {
            result.success -> ResponseEntity.ok(response)
            result.traceback != null -> ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response)
            else -> ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response)
        }
    }

    @PostMapping("/validate-config")
    fun validateConfig(@RequestBody request: Map<String, String>): ResponseEntity<Map<String, Any?>> {
        val configCode = request["configCode"]
        if (configCode.isNullOrBlank()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
                mapOf("success" to false, "error" to "Missing configCode")
            )
        }

        val result = mutableMapOf<String, Any?>(
            "success" to true,
            "sdk" to "android",
            "sdkVersion" to "unknown",
            "warnings" to emptyList<String>(),
            "resolvedOptions" to emptyMap<String, Any>(),
            "recognizedKeys" to emptyList<String>(),
            "ignoredKeys" to emptyList<String>()
        )

        try {
            val shell = groovy.lang.GroovyShell()
            shell.evaluate(configCode)
            result["initSucceeded"] = true
        } catch (e: Exception) {
            result["initSucceeded"] = false
            result["error"] = e.message
        }

        return ResponseEntity.ok(result)
    }

    @GetMapping("/introspect")
    fun introspect(): ResponseEntity<Map<String, Any>> {
        val options = listOf(
            mapOf("key" to "dsn", "canonicalKey" to "dsn", "type" to "string", "required" to true, "default" to null, "description" to "Data Source Name"),
            mapOf("key" to "debug", "canonicalKey" to "debug", "type" to "boolean", "required" to false, "default" to false, "description" to "Enable debug mode"),
            mapOf("key" to "release", "canonicalKey" to "release", "type" to "string", "required" to false, "default" to null, "description" to "Release version"),
            mapOf("key" to "environment", "canonicalKey" to "environment", "type" to "string", "required" to false, "default" to null, "description" to "Environment name"),
            mapOf("key" to "sampleRate", "canonicalKey" to "sampleRate", "type" to "float", "required" to false, "default" to 1.0, "description" to "Error sample rate"),
            mapOf("key" to "tracesSampleRate", "canonicalKey" to "tracesSampleRate", "type" to "float", "required" to false, "default" to null, "description" to "Traces sample rate"),
            mapOf("key" to "beforeSend", "canonicalKey" to "beforeSend", "type" to "function", "required" to false, "default" to null, "description" to "Hook before sending event"),
            mapOf("key" to "maxBreadcrumbs", "canonicalKey" to "maxBreadcrumbs", "type" to "integer", "required" to false, "default" to 100, "description" to "Max breadcrumbs"),
            mapOf("key" to "attachStacktrace", "canonicalKey" to "attachStacktrace", "type" to "boolean", "required" to false, "default" to true, "description" to "Attach stacktrace to messages"),
            mapOf("key" to "sendDefaultPii", "canonicalKey" to "sendDefaultPii", "type" to "boolean", "required" to false, "default" to false, "description" to "Send default PII"),
            mapOf("key" to "enableAutoSessionTracking", "canonicalKey" to "enableAutoSessionTracking", "type" to "boolean", "required" to false, "default" to true, "description" to "Enable auto session tracking"),
            mapOf("key" to "sessionTrackingIntervalMillis", "canonicalKey" to "sessionTrackingIntervalMillis", "type" to "integer", "required" to false, "default" to 30000, "description" to "Session tracking interval"),
            mapOf("key" to "anrEnabled", "canonicalKey" to "anrEnabled", "type" to "boolean", "required" to false, "default" to true, "description" to "Enable ANR detection"),
            mapOf("key" to "anrTimeoutIntervalMillis", "canonicalKey" to "anrTimeoutIntervalMillis", "type" to "integer", "required" to false, "default" to 5000, "description" to "ANR timeout interval")
        )

        return ResponseEntity.ok(mapOf(
            "sdk" to "android",
            "sdkVersion" to "unknown",
            "sdkPackage" to "io.sentry:sentry-android",
            "source" to "manifest",
            "options" to options,
            "timestamp" to java.time.Instant.now().toString()
        ))
    }

    @GetMapping("/health")
    fun health(): ResponseEntity<Map<String, String>> {
        return ResponseEntity.ok(
            mapOf(
                "status" to "healthy",
                "sdk" to "android"
            )
        )
    }
}
